packages/grafana-runtime/src/measurement/query.ts

```diff
@@ -0,0 +1,72 @@
+import {
+  DataQueryResponse,
+  isLiveChannelMessageEvent,
+  isLiveChannelStatusEvent,
+  isValidLiveChannelAddress,
+  LiveChannelAddress,
+  LoadingState,
+} from '@grafana/data';
+import { LiveMeasurements, MeasurementsQuery } from './types';
+import { getGrafanaLiveSrv } from '../services/live';
+
+import { Observable, of } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+export function getLiveMeasurements(addr: LiveChannelAddress): LiveMeasurements | undefined {
+  if (!isValidLiveChannelAddress(addr)) {
+    return undefined;
+  }
+
+  const live = getGrafanaLiveSrv();
+  if (!live) {
+    return undefined;
+  }
+
+  const channel = live.getChannel<LiveMeasurements>(addr);
+  const getController = channel?.config?.getController;
+  return getController ? getController() : undefined;
+}
+
+/**
+ * When you know the stream will be managed measurments
+ */
+export function getLiveMeasurmentsObserver(
+  addr: LiveChannelAddress,
+  requestId: string,
+  query?: MeasurementsQuery
+): Observable<DataQueryResponse> {
+  const rsp: DataQueryResponse = { data: [] };
+  if (!addr || !addr.path) {
+    return of(rsp); // Address not configured yet
+  }
+
+  const live = getGrafanaLiveSrv();
+  if (!live) {
+    // This will only happen with the feature flag is not enabled
+    rsp.error = { message: 'Grafana live is not initalized' };
+    return of(rsp);
+  }
+
+  rsp.key = requestId;
+  return live
+    .getChannel<LiveMeasurements>(addr)
+    .getStream()
+    .pipe(
+      map(evt => {
+        if (isLiveChannelMessageEvent(evt)) {
+          rsp.data = evt.message.getData(query);
+          if (!rsp.data.length) {
+            // ?? skip when data is empty ???
+          }
+          delete rsp.error;
+          rsp.state = LoadingState.Streaming;
+        } else if (isLiveChannelStatusEvent(evt)) {
+          if (evt.error != null) {
+            rsp.error = rsp.error;
+            rsp.state = LoadingState.Error;
+          }
+        }
+        return { ...rsp }; // send event on all status messages
+      })
+    );
```

Comment:
nit: this looks fine but everytime you introduce an if statement in RxJs operator you should always ask yourself if that operator can be divided into 2 streams using `filter` and then merge the streams in the return.

Reply:
why is that preferable? Seems more complex & inefficient? Can understand it if you after the filtering need to do different async stream processing

Reply:
I can see you point Torkel and maybe this is not the perfect use case because of efficiency and because we don't do anything combined. But usually the streams tend to grow as all other code and then I've found that keeping the logic outside the map functions is easier to read and easier to refactor. 

I was just trying to share some of the insights I found working with larger RxJs implementations, I hold no strong opinions about this.